pinterest/teletraan
# Store ping timestamp per host separately.

Problem:
----------
Host table keeps inventory of all the hosts registered to Teletraan.
1. It keeps timestamp of all the registered hosts -> used by AgentJanitor to cleanup unreachable or dead hosts.
2. It also keep mapping between host and group. (For example: helloworlddummyservice-server-dev-0a01def3 <> CMP, helloworlddummyservice-server-dev-0a01def3 <> mcrouter-prod-sidecar, ... etc.) -> used in UI/API to populate hosts which are part of the specific environment.

Due to this timestamp needs to be updated per host <-> group mapping for a given host instead of just per host.

Fix:
----
1. Extend existing hosts_and_agents table to keep per host data (specifically timestamp). 
With this change, 
- last_update field of hosts table will not be updated on every ping. Instead last_update in hosts_and_agents table will be updated on every ping once per host id.
- host <-> group mapping will be inserted once for new host and then after for any addition/deletion of host <-> group mapping.
This potentially reduces inserts by 7 folds per host per ping.
(Note: fields from hosts table are not removed for backward compatibility)

2. Improved on AgentJanitor logic by removing group level processing as that loop iterate through all the groups in the system to find unreachable hosts -> which I think is a processing waste. 
AgentJanitor will continue to go through all the stale hosts in the system with added asg info in hosts_and_agents table so as to pull host launch latency threshold config from ASG for a given stale host.

--
Other fields - state, can_retire - can also be moved to host_and_agents table as those are also per host. However, as that requires substantial changes at API level and also client side,  those changes are not included in the scope.


Tests:
---
1. Verify host_name, host_id, ip, asg, create_date, last_update, agent_version of a host in hosts_and_agents table.
2. Verify last_update field is not updated in hosts table on every ping if there is no change in host <-> group mapping.
3. Verify that new deploy groups are updated in hosts table.
4. Verify that stale deploy groups are removed from hosts table.
2. Terminate host part of ASG and verify it's cleaned up.
2. Remove deployd from host part of ASG and verify it's marked unreachable.
3. Terminate standalone hosts and verify it's cleaned up.
4. Remove deployd from standalone host and verify it's marked unreachable.
